### PR TITLE
fix: RewardBallotWinners determinism

### PR DESCRIPTION
**Pull request by robert-zaremba**
## Description

Fixes RC1 Canon consensus failure.

Bug was in: https://github.com/umee-network/umee/pull/1433/files#diff-971aeedb004816fd5770d86dfa2d9ee39c5e53d7802773dd9e81cbc2cf890f62L60

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] added appropriate labels to the PR
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

**Comment by robert-zaremba** [comment]
> approved +1 was this introduced in #1391 or was this in the oracle previously / should we have seen that in the latest beta?

No. #1381 didn't touch `RewardBallotWinners`. Basically, the bug is that we iterate over a map for `k.distrKeeper.AllocateTokensToValidator`  to send distributions in an AVL tree. I will make it more clear in few secs

File: x/oracle/keeper/reward.go [hunk 1]
```diff
@@ -57,6 +57,7 @@ func (k Keeper) RewardBallotWinners(
 
 	// distribute rewards
 	var distributedReward sdk.Coins
+
 	for _, winner := range ballotWinners {
```
Comment (robert-zaremba) [inline x/oracle/keeper/reward.go:61]: here we had a bug, we need to iterate over slice, because in the loop we are updating distribution state.

**Comment by toteki** [comment]
Map bugs are elusive - they'll only reveal themselves if the underlying slices were allocated differently, and the conditions when that will appear in practice are undefined.

Probably didn't appear in betas due to the small number of validators, operating under similar conditions

**Comment by RafilxTenfen** [comment]
ohhh, so in the sum part it didn't matter that much, because sum the whole amount in map or slice is the same thing, 
but when we iterate again [here](https://github.com/umee-network/umee/blob/547843e1ed644f2b083feb8875abc8e831573726/x/oracle/keeper/reward.go#L60) to send rewards, it matters

**Comment by robert-zaremba** [comment]
@RafilxTenfen correct.

**Pull request merged**